File: 301-400/381-390/381-insertDeleteGetRandom2/insertDeleteGetRandom2.cpp
class RandomizedCollection {
public:
    vector<int> nums;
    unordered_multimap<int, int> m;

    /** Initialize your data structure here. */
    RandomizedCollection() {
    }

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        nums.emplace_back(val);
        m.insert(make_pair(val, nums.size() - 1));
        return true;
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        auto iter = m.find(val);
        if (iter == m.end())
            return false;

        int last = nums.back();
        auto range = m.equal_range(last);
        for (auto range_iter = range.first; range_iter != range.second; ++range_iter)
            if (range_iter->second == nums.size() - 1)
                range_iter->second = iter->second;

        nums[iter->second] = last;
        nums.pop_back();
        m.erase(iter);
        return true;
    }

    /** Get a random element from the set. */
    int getRandom() {
        return nums[rand() % nums.size()];
    }
};

/**
 * Your RandomizedCollection object will be instantiated and called as such:
 * RandomizedCollection* obj = new RandomizedCollection();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 301-400/381-390/381-insertDeleteGetRandom2/insertDeleteGetRandom2.py
import random
import collections

class RandomizedCollection(object):
    def __init__(self):
        self.vals, self.idxs = [], collections.defaultdict(set)


    def insert(self, val):
        self.vals.append(val)
        self.idxs[val].add(len(self.vals) - 1)
        return len(self.idxs[val]) == 1


    def remove(self, val):
        if self.idxs[val]:
            out, ins = self.idxs[val].pop(), self.vals[-1]
            self.vals[out] = ins
            if self.idxs[ins]:
                self.idxs[ins].add(out)
                self.idxs[ins].discard(len(self.vals) - 1)
            self.vals.pop()
            return True
        return False

    def getRandom(self):
        return random.choice(self.vals)